File: solutions/vbnet/circular-buffer/1/CircularBuffer.vb
Imports System
Imports System.Collections.Generic

Public Class CircularBuffer(Of T)
    Inherits Queue(Of T)

    Public ReadOnly Property Capacity As Integer

    Public Sub New(ByVal capacity As Integer)
        Me.Capacity = capacity
    End Sub

    Public Function Read() As T
        If Count = 0 Then
            Throw New InvalidOperationException("buffer is empty.")
        End If
        Dim item = Dequeue()
        Return item
    End Function

    Public Sub Write(ByVal value As T)
        If Count = Capacity Then
            Throw New InvalidOperationException("buffer is full.")
        End If
        Enqueue(value)
    End Sub

    Public Sub Overwrite(ByVal value As T)
        If Count = Capacity Then
            Dequeue()
        End If
        Enqueue(value)
    End Sub
End Class

File: solutions/delphi/circular-buffer/1/circular_buffer.pas
unit uCircularBuffer;

interface

uses
  System.Generics.Collections, System.StrUtils, System.SysUtils;

type
  EInvalidOpException = class(exception);

  ICircularBuffer<T> = interface
  ['{31315AD8-4E00-4722-BF57-F220B3F2451C}']
    procedure Write(AValue:T);
    function Read:T;
    procedure Clear;
    procedure OverWrite(AValue:T);
  end;

  TCircularBuffer<T> = class(TInterfacedObject, ICircularBuffer<T>)
  private
    queue:TQueue<T>;
  public
    procedure Write(AValue:T);
    function Read:T;
    procedure Clear;
    procedure OverWrite(AValue:T);
    constructor Create(const ALen:integer);
    destructor Destroy;
  end;

implementation


{ TCircularBuffer<T> }

procedure TCircularBuffer<T>.Clear;
begin
  queue.Clear;
end;

constructor TCircularBuffer<T>.Create(const ALen: integer);
begin
  queue := TQueue<T>.Create;
  queue.Capacity := ALen;
end;

destructor TCircularBuffer<T>.Destroy;
begin
  FreeAndNil(queue);
end;

procedure TCircularBuffer<T>.OverWrite(AValue: T);
begin
  if (queue.Count = queue.Capacity) then
    queue.Dequeue;
  queue.Enqueue(AValue);
end;

function TCircularBuffer<T>.Read: T;
begin
  if queue.Count = 0 then
    raise EInvalidOpException.Create('empty buffer');
  Result := queue.Dequeue;
end;

procedure TCircularBuffer<T>.Write(AValue: T);
begin
  if (queue.Count = queue.Capacity) then
    raise EInvalidOpException.Create('full buffer');
  queue.Enqueue(AValue);
end;

end.